Implement 2D instance uniforms

paddy-exe opened the pull request:
Adds support for 2d instance shader parameters. Cherrypicked the original PR as first commit.

Supersedes https://github.com/godotengine/godot/pull/97058
Fixes https://github.com/godotengine/godot/issues/62943

Also works with `use_parent_material`:

<img width="1433" alt="image" src="https://github.com/user-attachments/assets/5cfa2167-5148-40be-8584-56d2f0f3e23e">

Testing project (taken from original PR):
[test_pr_97058.zip](https://github.com/user-attachments/files/18312291/test_pr_97058.zip)

KoBeWi committed:
Implement 2D instance shader parameters

Co-authored-by: kobewi <kobewi4e@gmail.com>
Co-authored-by: yesfish <huwpascoe@users.noreply.github.com>
Co-authored-by: Álex Román Núñez <eirexe123@gmail.com>

akien-mga commented:
Can you link that original PR? Is this new PR meant to supersede it, or it's a follow-up and KoBeWi's PR needs to be merged first?

paddy-exe commented:
> Can you link that original PR? Is this new PR meant to supersede it, or it's a follow-up and KoBeWi's PR needs to be merged first?

It's the follow-up PR to https://github.com/godotengine/godot/pull/97058 but it is in VERY early stages. I am trying to wrap my head around the code. Clay gave me instructions on how it could be done for the compatability renderer. All the rest is already done

paddy-exe commented on drivers/gles3/rasterizer_canvas_gles3.h:
```diff
@@ -219,7 +219,7 @@ class RasterizerCanvasGLES3 : public RendererCanvasRender {
 				float dst_rect[4];
 				float src_rect[4];
-				float pad[2];
+				uint32_t instance_uniforms_ofs;
```
Not sure if I am setting this in the right place...

clayjohn replied:
No, this isn't the right place. But it makes me realize that more changes are needed than I described before, sorry. 

We need to move `specular_shininess` into a regular uniform. That will be easy because it is something that is the same for the entire draw call. Then `instance_uniforms_ofs` should go in the spot where `specular_shininess` used to be

clayjohn commented:
I started writing up what to do, then I realized it was less work for me to just show you, so here is a patch that gets 2D instance uniforms working in 2D.

[0001-Finish-GLES3-support.patch](https://github.com/user-attachments/files/17878926/0001-Finish-GLES3-support.patch)

The final piece before this is ready is getting it to work with the "use_parent_material" property. CanvasItems often inherit the material from their parent. Currently it mostly works, but the editor doesn't expose the instance property for updating so you can't set the instance uniform in the editor.

Chaosus commented on doc/classes/GeometryInstance3D.xml:
```diff
@@ -25,7 +25,6 @@
 				Set the value of a shader uniform for this instance only ([url=$DOCS_URL/tutorials/shaders/shader_reference/shading_language.html#per-instance-uniforms]per-instance uniform[/url]). See also [method ShaderMaterial.set_shader_parameter] to assign a uniform on all instances using the same [ShaderMaterial].
 				[b]Note:[/b] For a shader uniform to be assignable on a per-instance basis, it [i]must[/i] be defined with [code]instance uniform ...[/code] rather than [code]uniform ...[/code] in the shader code.
 				[b]Note:[/b] [param name] is case-sensitive and must match the name of the uniform in the code exactly (not the capitalized name in the inspector).
-				[b]Note:[/b] Per-instance shader uniforms are currently only available in 3D, so there is no 2D equivalent of this method.
```
Worth mentioned that is only available on `3D`/`2D` but not for `Fog`, `Sky `or `Particles`.

paddy-exe replied:
Amended, thank you!

paddy-exe commented:
> I started writing up what to do, then I realized it was less work for me to just show you, so here is a patch that gets 2D instance uniforms working in 2D.
> 
> [0001-Finish-GLES3-support.patch](https://github.com/user-attachments/files/17878926/0001-Finish-GLES3-support.patch)
> 
> The final piece before this is ready is getting it to work with the "use_parent_material" property. CanvasItems often inherit the material from their parent. Currently it mostly works, but the editor doesn't expose the instance property for updating so you can't set the instance uniform in the editor.

Thank you! I applied your patch except for this part as I couldn't figure out where this goes:

```cpp
material_storage->shaders.canvas_shader.version_set_uniform(CanvasShaderGLES3::SPECULAR_SHININESS_IN, state.canvas_instance_batches[i].specular_shininess, shader_version, variant, specialization);
```

paddy-exe commented:
Updated with all feedback. Once the code changes are approved I will squash them so we have one commit at the end. The current two commits are for easier checking of the changes I made versus the previous authors made.

Mickeon reviewed (commented):
I would also suggest touching on the descriptions of the equivalent methods for instance geometry, so that they're consistent with the ones added in this PR (which are slightly better formatted). But it doesn't have to be done in this PR.

clayjohn commented on scene/main/canvas_item.cpp:
```diff
@@ -587,7 +587,14 @@ bool CanvasItem::_get(const StringName &p_name, Variant &r_ret) const {
 	List<PropertyInfo> pinfo;
-	RS::get_singleton()->canvas_item_get_instance_shader_parameter_list(get_canvas_item(), &pinfo);
+	if (use_parent_material) {
+		Node *parent = get_parent();
+		CanvasItem *parent_canvas_item = Object::cast_to<CanvasItem>(parent);
+		RS::get_singleton()->canvas_item_get_instance_shader_parameter_list(parent_canvas_item->get_canvas_item(), &pinfo);
```
I don't think this is correct. What if the parent has `use_parent_material` enabled as well?

paddy-exe replied:
In that case they will both use the material of the parent:
<img width="1446" alt="image" src="https://github.com/user-attachments/assets/6dafb4ea-2714-4768-925f-6e84ba123d2d">

<img width="1468" alt="image" src="https://github.com/user-attachments/assets/9526263a-db30-40c2-ada6-47dbfe11f98b">

<img width="1439" alt="image" src="https://github.com/user-attachments/assets/80499fa3-20b3-49b5-8f89-ee5e8979cc98">

paddy-exe replied:
I am confused now. If I have the following structure
```
Node A
-- Node B
----Node C
```
If Node A has the instance shader applied, Node B recognizes it if `use_parent_material` is enabled and will show the params in the inspector where they can be altered independently from the params of Node A. If Node C now also has `use_parent_material` enabled then it will inherit the material from Node B which itself inherits it from Node A. And the inspector still pops up for both Node B and Node C since they have the `use_parent_material` enabled. If you disable it, the inspector params disappear again.

clayjohn replied:
This is how it should be yes, but that isn't how it works right now. Right now both Node B and Node C will inherit the material, but only Node B will be able to modify the parameters in the inspector. 

In this example Icon, Icon2, and Icon3 all have "use_parent_material" enabled. The material is passed down to them, but the uniform set in the inspector is ignored. Then for icon3 the inspector option to set the shader parameter doesn't even appear. 

<img width="1435" alt="Screenshot 2024-11-27 at 12 49 22 PM" src="https://github.com/user-attachments/assets/50fff287-111d-4d9b-833a-44cc0c7619d9">

ettiSurreal commented on doc/classes/GeometryInstance3D.xml:
```diff
@@ -25,7 +25,7 @@
 				[b]Note:[/b] For a shader uniform to be assignable on a per-instance basis, it [i]must[/i] be defined with [code]instance uniform ...[/code] rather than [code]uniform ...[/code] in the shader code.
 				[b]Note:[/b] [param name] is case-sensitive and must match the name of the uniform in the code exactly (not the capitalized name in the inspector).
-				[b]Note:[/b] Per-instance shader uniforms are currently only available in 3D, so there is no 2D equivalent of this method.
+				[b]Note:[/b] Per-instance shader uniforms are only available in 3D and 2D but not for Fog, Sky or Particles.
```
Wouldn't it be better to specify Spatial and CanvasItem?

tetrapod00 replied:
```suggestion
				[b]Note:[/b] Per-instance shader uniforms are only available in Spatial and CanvasItem shaders, but not Fog, Sky or Particles shaders.
```

Mickeon replied:
The descriptions you wrote for the new methods in this PR correspond to the existing descriptions for the instance geometry methods, but are ever-so-slightly altered for the best.
For example, compared to `instance_geometry_get_shader_parameter_list`, `canvas_item_get_instance_shader_parameter_list` splits the description in two lines and omits the _"Equivalent to [method GeometryInstance3D.get_instance_shader_parameter]."_ because it does not make sense.
In hindsight though, there's no need to do this now.

Calinou reviewed (approved):
Tested locally, it works as expected in all rendering methods.

**Testing project:** [test_pr_99230.zip](https://github.com/user-attachments/files/17934918/test_pr_99230.zip)

AThousandShips commented on doc/classes/GeometryInstance3D.xml:
```diff
@@ -25,7 +25,7 @@
 				[b]Note:[/b] For a shader uniform to be assignable on a per-instance basis, it [i]must[/i] be defined with [code]instance uniform ...[/code] rather than [code]uniform ...[/code] in the shader code.
 				[b]Note:[/b] [param name] is case-sensitive and must match the name of the uniform in the code exactly (not the capitalized name in the inspector).
-				[b]Note:[/b] Per-instance shader uniforms are currently only available in 3D, so there is no 2D equivalent of this method.
+				[b]Note:[/b] Per-instance shader uniforms are only available in Spatial and CanvasItem shaders, but not for Fog, Sky or Particles shaders.
```
```suggestion
				[b]Note:[/b] Per-instance shader uniforms are only available in Spatial and CanvasItem shaders, but not for Fog, Sky, or Particles shaders.
```

AThousandShips commented on servers/rendering/renderer_canvas_cull.cpp:
```diff
@@ -44,6 +44,29 @@
 // while not making lines appear too soft.
 const static float FEATHER_SIZE = 1.25f;
 
+static RendererCanvasCull *_canvas_cull_singleton{};
```
```suggestion
static RendererCanvasCull *_canvas_cull_singleton = nullptr;
```

AThousandShips commented on servers/rendering/renderer_canvas_cull.cpp:
```diff
@@ -1884,6 +1911,37 @@ void RendererCanvasCull::canvas_item_set_use_parent_material(RID p_item, bool p_
+void RendererCanvasCull::canvas_item_get_instance_shader_parameter_list(RID p_item, List<PropertyInfo> *p_parameters) const {
+	const Item *item = const_cast<RendererCanvasCull *>(this)->canvas_item_owner.get_or_null(p_item);
+	ERR_FAIL_NULL(item);
+	ERR_FAIL_NULL(p_parameters);
```
I'd suggest checking this first before doing any lookups

paddy-exe replied:
@AThousandShips Would you mind checking again if I have done this correctly?

AThousandShips commented on servers/rendering/renderer_scene_cull.cpp:
```diff
@@ -1701,18 +1664,11 @@ uint32_t RendererSceneCull::get_pipeline_compilations(RS::PipelineSource p_sourc
 void RendererSceneCull::instance_geometry_get_shader_parameter_list(RID p_instance, List<PropertyInfo> *p_parameters) const {
 	const Instance *instance = const_cast<RendererSceneCull *>(this)->instance_owner.get_or_null(p_instance);
 	ERR_FAIL_NULL(instance);
+	ERR_FAIL_NULL(p_parameters);
```
Same here

huwpascoe commented:
I'd like to test this, is it possible to get those conflicts resolved?

paddy-exe commented:
@huwpascoe Conflicts resolved. 

Be aware that currently the editor doesn't show the parent material properties when `use_parent_material` is enabled. Still working on that part.

huwpascoe commented:
> Be aware that currently the editor doesn't show the parent material properties when `use_parent_material` is enabled. Still working on that part.

I didn't even know that was a feature....well I remember enough about my own work to fix that!

@paddy-exe Here is a patch that implements the parent materials.
[0001-implement-use_parent_material.patch](https://github.com/user-attachments/files/18159048/0001-implement-use_parent_material.patch)

paddy-exe commented:
Alright, patch applied. Works great. Thank you for the assistance with this @huwpascoe. Squashed into one commit. Adjusted the code to AThousandShip's codestyle feedback.

AThousandShips commented on servers/rendering/renderer_canvas_cull.cpp:
```diff
@@ -1875,13 +1901,46 @@ void RendererCanvasCull::canvas_item_set_material(RID p_item, RID p_material) {
+}
+
+void RendererCanvasCull::canvas_item_get_instance_shader_parameter_list(RID p_item, List<PropertyInfo> *p_parameters) const {
+	ERR_FAIL_NULL(p_item);
+	ERR_FAIL_NULL(p_parameters);
+	const Item *item = const_cast<RendererCanvasCull *>(this)->canvas_item_owner.get_or_null(p_item);
```
```suggestion
	ERR_FAIL_NULL(p_parameters);
	const Item *item = const_cast<RendererCanvasCull *>(this)->canvas_item_owner.get_or_null(p_item);
	ERR_FAIL_NULL(item);
```
Meant just specifically the `p_parameters` so the null check should be moved down and check `item` as the `RID` can't be checked

AThousandShips replied:
Should probably still check for null for `item` and `instance` in each of these

paddy-exe replied:
Hmm... interestingly if I add a null check for instance and item the engine crashes :/

AThousandShips replied:
That's weird can you show your code?

AThousandShips commented on servers/rendering/renderer_scene_cull.cpp:
```diff
@@ -1699,20 +1662,12 @@ uint32_t RendererSceneCull::get_pipeline_compilations(RS::PipelineSource p_sourc
 void RendererSceneCull::instance_geometry_get_shader_parameter_list(RID p_instance, List<PropertyInfo> *p_parameters) const {
+	ERR_FAIL_NULL(p_parameters);
 	const Instance *instance = instance_owner.get_or_null(p_instance);
-	ERR_FAIL_NULL(instance);
```
I.e. restore this, as it is not guaranteed to be valid

paddy-exe replied:
Just to clarify: You mean remove it or add it?

AThousandShips replied:
Put it back the way it was

paddy-exe replied:
Alright. Thanks for clarifying. Should be done now I think

AThousandShips commented on servers/rendering/renderer_canvas_cull.cpp:
```diff
@@ -1875,13 +1901,44 @@ void RendererCanvasCull::canvas_item_set_material(RID p_item, RID p_material) {
+
+void RendererCanvasCull::canvas_item_get_instance_shader_parameter_list(RID p_item, List<PropertyInfo> *p_parameters) const {
+	ERR_FAIL_NULL(p_parameters);
+	const Item *item = const_cast<RendererCanvasCull *>(this)->canvas_item_owner.get_or_null(p_item);
```
```suggestion
	const Item *item = const_cast<RendererCanvasCull *>(this)->canvas_item_owner.get_or_null(p_item);
	ERR_FAIL_NULL(item);
```

clayjohn reviewed (approved):
Looks good to me. Tested locally and confirmed it is still working including with `use_parent_material`. I think it is ready to merge to get broader testing!!

Repiteo commented:
Thanks!
